cmd/opampsupervisor/e2e_test.go

```diff
@@ -1262,6 +1262,148 @@ func TestSupervisorWritesAgentFilesToStorageDir(t *testing.T) {
+	agentFileCopyPath := filepath.Join(tmpDir, agentFileName)
+
+	// Upgrading will overwrite the agent binary, so we'll copy to a new path to not affect other tests
+	copyFile(t, agentFilePath, agentFileCopyPath)
+
+	agentIDChan := make(chan []byte, 1)
+	agentDescriptionChan := make(chan *protobufs.AgentDescription, 1)
+	packageStatusesChan := make(chan *protobufs.PackageStatuses, 2)
+
+	server := newOpAMPServer(
+		t,
+		defaultConnectingHandler,
+		server.ConnectionCallbacksStruct{
+			OnMessageFunc: func(_ context.Context, _ types.Connection, message *protobufs.AgentToServer) *protobufs.ServerToAgent {
+				select {
+				case agentIDChan <- message.InstanceUid:
+				default:
+				}
+
+				if message.AgentDescription != nil {
+					select {
+					case agentDescriptionChan <- message.AgentDescription:
+					default:
+					}
+				}
+
+				if message.PackageStatuses != nil {
+					select {
+					case packageStatusesChan <- message.PackageStatuses:
+					default:
+					}
+				}
+
+				return &protobufs.ServerToAgent{}
+			},
+		},
+	)
+
+	s := newSupervisor(t, "upgrade", map[string]string{
+		"url":         server.addr,
+		"storage_dir": storageDir,
+		"agent_path":  agentFileCopyPath,
+	})
+
+	require.Nil(t, s.Start())
+	defer s.Shutdown()
+
+	waitForSupervisorConnection(server.supervisorConnected, true)
+
+	t.Logf("Supervisor connected")
+
+	agentVersion := "0.110.0"
+	agentName := fmt.Sprintf("otelcol-contrib_0.110.0_%s_%s.tar.gz", runtime.GOOS, runtime.GOARCH)
+	agentURL := fmt.Sprintf("https://github.com/open-telemetry/opentelemetry-collector-releases/releases/download/v0.110.0/%s", agentName)
+	agentSigURL := fmt.Sprintf("https://github.com/open-telemetry/opentelemetry-collector-releases/releases/download/v0.110.0/%s.sig", agentName)
+	agentCertURL := fmt.Sprintf("https://github.com/open-telemetry/opentelemetry-collector-releases/releases/download/v0.110.0/%s.pem", agentName)
+
+	agentHash := getHTTPBodyHash(t, agentURL)
+	cert := getHTTPBodyContents(t, agentCertURL)
+	sig := getHTTPBodyContents(t, agentSigURL)
+
+	signatureField := bytes.Join([][]byte{cert, sig}, []byte(" "))
+
+	<-packageStatusesChan
+	<-agentDescriptionChan
+	agentID := <-agentIDChan
+	server.sendToSupervisor(&protobufs.ServerToAgent{
+		InstanceUid: agentID,
+		PackagesAvailable: &protobufs.PackagesAvailable{
+			Packages: map[string]*protobufs.PackageAvailable{
+				"": {
+					Type:    protobufs.PackageType_PackageType_TopLevel,
+					Version: "v" + agentVersion,
+					Hash:    []byte{0x01, 0x02},
+					File: &protobufs.DownloadableFile{
+						DownloadUrl: agentURL,
+						ContentHash: agentHash,
+						Signature:   signatureField,
+					},
+				},
+			},
+			AllPackagesHash: []byte{0x03, 0x04},
+		},
+	})
+
+	// Wait for new package statuses
+	ps := <-packageStatusesChan
+	require.Equal(t, &protobufs.PackageStatuses{
+		Packages: map[string]*protobufs.PackageStatus{
+			"": {
+				// TODO: Should initital version be filled in?
```

Comment:
Yeah, I think that makes sense. I think the mental hurdle I'm having here is the sort of disconnect between the artifact the agent needs to end up with and what is available from the releases.

Because the release artifact is a tarball, but the agent executable is not, it means that the artifact offered by the server and the one the agent ends up with have completely different hashes. This initial status is kind of a weird edge case.